pass performance param

grkhr opened the pull request.

grkhr commented:
I have read the CLA Document and I hereby sign the CLA

bh2smith commented on .gitignore:
```diff
@@ -6,3 +6,4 @@ _version.py
 .idea/
 venv/
 tmp/
+.vscode/settings.json
```
note that this can also be set in your global .git configuration


with this

```
[core]
	excludesfile = ~/.gitignore_global
```
and 

```
➜  ~ cat .gitignore_global 
venv/
.DS_Store
.idea/
.vscode/
```
```suggestion
.vscode/
```

however, I will allow it since `.idea/` is somehow already in here.

bh2smith commented on dune_client/client.py:
```diff
@@ -71,14 +71,15 @@ def _post(self, route: str, params: Any) -> Any:
                     p.key: p.to_dict()["value"] for p in query.parameters()
-                }
+                },
+                "performance": performance,
```
If we put performance as an attribute of the client, then this would be one of the very few lines of code that needs to change here. Like so,

```suggestion
                "performance": self.performance,
```


The the constructor would be

```py
def __init__(self, api_key: str, performance: str = "medium"):
    self.performance = performance
    ...
```

grkhr replied:
I was thinking about this too, but in our case we initialize the client once and then we want to execute a lot of queries on different cluster sizes.

e.g.
```
dc = DuneClient(...)

dc.refresh(123, 'medium')
dc.refresh(456, 'large')
...
```

grkhr replied:
That's a good idea! So don't merge, ill add some changes

Also, I've found that there is no latest results endpoint implementation. I'm gonna rename this PR and add these changes as well.

bh2smith reviewed (commented):
Overall I think these changes look really good. Thanks very much for you contribution!

However, I think that we could get away with fewer default values set (like private/internal methods can expect that these values will be set), but it doesn't have to be changed.

Specifically wondering if it might make sense to put performance attribute on the client itself (i.e. be set one time for an instance of a single client) so that we don't have to pass it into every function. then there would be only one place that its actually referenced (inside of `execute`). Then end consumers could declare `fast_execution_client` and `medium_execution_client`. What do you think about this? It would result in much simpler code design and fewer redundant changes

msf reviewed (approved):
looks good to me.
I think it makes sense to optionally specific performance per `request()` request.. 
Having an instance level default is a nice to have, so that one can always execute on a higher tier if one wants.

grkhr commented:
@bh2smith 

- Passed "performance" param to __init__, cleaned up some mess with that param passing through
- Implemented "latest result" endpoint ([api doc](https://dune.com/docs/api/api-reference/latest_results/))

msf commented on dune_client/client_async.py:
```diff
@@ -134,6 +141,14 @@ async def get_result(self, job_id: str) -> ResultsResponse:
             raise DuneError(response_json, "ResultsResponse", err) from err
+        
+    async def get_latest_result(self, query: Query) -> ResultsResponse:
+        """GET results from Dune API for `query_id`"""
```
`GET the latest results for a query_id without having to execute the query again.`
you can also include the link to the API docs if you'd like

grkhr commented on dune_client/client.py:
```diff
@@ -121,6 +124,14 @@ def get_result_csv(self, job_id: str) -> ExecutionResultCSV:
         response.raise_for_status()
         return ExecutionResultCSV(data=BytesIO(response.content))
+    
+    def get_latest_result(self, query: Query) -> ResultsResponse:
```
I vote for `get_latest_query_result` (btw without plural according to other function's naming)

bh2smith replied:
Personally, I prefer the `get_latest_result` since "query" is already an explicit function parameter and its shorter, but sufficiently descriptive term. However, it appears I am already outvoted so I wouldn't block this from being merged with the long name.

grkhr replied:
Fair enough, let's go with `get_latest_result`

bh2smith commented on dune_client/client_async.py:
```diff
@@ -135,6 +145,14 @@ async def get_result(self, job_id: str) -> ResultsResponse:
         except KeyError as err:
             raise DuneError(response_json, "ResultsResponse", err) from err
 
+    async def get_latest_result(self, query: Query) -> ResultsResponse:
+        """GET latest results from Dune API for `query_id`"""
+        response_json = await self._get(url=f"query/{query.query_id}/results")
```
It looks like Query object isn't even necessary here (just the ID). I am trying to understand how to fetch the latest result for a given set or parameters. Does not seem to be part of this, but I am pretty sure this works in the interface.

grkhr replied:
Thanks for noticing that. I'll check the API behaviour in case of non-empty query params.

grkhr replied:
Committed the update. Not sure if it's ok, can you check @bh2smith?

grkhr replied:
In case if there is no result for specified set of params it will return 

`DuneError: Can't build ResultsResponse from {'error': 'not found: No execution found for the latest version of the given query'}`

bh2smith commented on dune_client/client.py:
```diff
@@ -132,8 +146,13 @@ def cancel_execution(self, job_id: str) -> bool:
+        self,
+        query: Query,
+        ping_frequency: int = 5,
+        performance: Union[str, None] = None,
```
Nit: Is there any reason you prefer `Union` over `Optional[str]` or even `str | None`? I guess we are still supporting python 3.7 so `str | None` wont work.

grkhr replied:
No reason tbh, will change to `Optional[str]`

grkhr replied:
Changed

bh2smith commented on dune_client/client.py:
```diff
@@ -128,9 +129,25 @@ def get_result_csv(self, job_id: str) -> ExecutionResultCSV:
-    def get_latest_result(self, query: Query) -> ResultsResponse:
-        """GET latest results from Dune API for `query_id`"""
-        response_json = self._get(route=f"query/{query.query_id}/results")
+    def get_latest_result(self, query: Union[Query, str, int]) -> ResultsResponse:
```
Can we update the doc string to describe better this Either or type we're using here.

grkhr replied:
done

bh2smith commented on dune_client/client.py:
```diff
@@ -128,9 +129,25 @@ def get_result_csv(self, job_id: str) -> ExecutionResultCSV:
+        https://dune.com/docs/api/api-reference/latest_results/
+        """
+        if isinstance(query, Query):
+            params = {
+                f"params.{p.key}": p.to_dict()["value"] for p in query.parameters()
+            }
```
interesting.... is this a new style?

grkhr replied:
yes, i was surprised too :)

https://dune.com/docs/api/api-reference/latest_results/#python

<img width="690" alt="image" src="https://github.com/cowprotocol/dune-client/assets/40689054/3b01c1c8-882e-4b6b-bbaa-d8d5f81b0872">

bh2smith replied:
Maybe you can add some tests for this route.

grkhr replied:
Yes, I've tested that locally. It's working.

I'll try to add some tests

bh2smith reviewed (approved):
Looks pretty!

grkhr commented:
So, if everything is ok, let's merge. And would be nice if you can push the new version to pypi so I can switch from custom git installation to pip.

bh2smith commented:
> So, if everything is ok, let's merge.

You wanna add some actual tests here? Or do you plan to follow up with some?

bh2smith commented:
Release published: https://github.com/cowprotocol/dune-client/releases/tag/v1.1.1


Should be on pypi within the next few minutes.
